src/dex/nerve/nerve-math.ts

```diff
@@ -0,0 +1,317 @@
+      tokenIndexFrom,
+      tokenIndexTo,
+      x,
+      xp,
+      blockTimestamp,
+    );
+
+    if (y === null) {
+      state.isValid = false;
+      return null;
+    }
+
+    // dy = xp[tokenIndexTo].sub(y).sub(1);
+    let dy = xp[tokenIndexTo] - y - ONE;
+
+    // dyFee = dy.mul(self.swapFee).div(FEE_DENOMINATOR);
+    const dyFee = (dy * state.swapFee) / this.FEE_DENOMINATOR;
+
+    // dy = dy.sub(dyFee).div(self.tokenPrecisionMultipliers[tokenIndexTo]);
+    dy = (dy - dyFee) / state.tokenPrecisionMultipliers[tokenIndexTo];
+    return { dy, dyFee };
+  }
+
+  calculateWithdrawOneToken(
+    state: PoolState,
+    tokenAmount: bigint,
+    tokenIndex: number,
+    blockTimeStamp: bigint,
+  ) {
+    const out = this._calculateWithdrawOneTokenDY(
+      state,
+      tokenIndex,
+      tokenAmount,
+      blockTimeStamp,
+    );
+    if (out === null) return null;
+    let { dy, newY } = out;
+
+    // uint256 dySwapFee = _xp(self)[tokenIndex].sub(newY)
+    //  .div(self.tokenPrecisionMultipliers[tokenIndex]).sub(dy)
+    const dySwapFee =
+      (this._xp(state)[tokenIndex] - newY) /
+        state.tokenPrecisionMultipliers[tokenIndex] -
+      dy;
+
+    // dy = dy.mul(FEE_DENOMINATOR.sub(calculateCurrentWithdrawFee(self, account)))
+    //  .div(FEE_DENOMINATOR);
+    dy =
+      (dy * (this.FEE_DENOMINATOR - this._calculateCurrentWithdrawFee(state))) /
+      this.FEE_DENOMINATOR;
+
+    return { dy, dyFee: dySwapFee };
+  }
+
+  swapUnderlying(state: PoolState) {}
+
+  protected _getNumTokens(state: PoolState) {
+    return biginterify(state.tokenPrecisionMultipliers.length);
+  }
+
+  protected _calculateWithdrawOneTokenDY(
+    state: PoolState,
+    tokenIndex: number,
+    tokenAmount: bigint,
+    blockTimestamp: bigint,
+  ) {
+    const numTokens = this._getNumTokens(state);
+    const xp = this._xp(state);
+    const v = {
+      d0: ZERO,
+      d1: ZERO,
+      newY: ZERO,
+      feePerToken: ZERO,
+      preciseA: ZERO,
+    };
+    v.preciseA = this._getAPrecise(state, blockTimestamp);
+    const d0 = this._getD(state, xp, v.preciseA);
+    if (d0 === null) return null;
+    else v.d0 = d0;
+
+    // v.d1 = v.d0.sub(tokenAmount.mul(v.d0).div(self.lpToken.totalSupply()));
+    v.d1 = v.d0 - (tokenAmount * v.d0) / state.lpToken_supply;
+    const newY = this._getYD(state, v.preciseA, tokenIndex, xp, v.d1);
+    if (newY === null) return null;
+    else v.newY = newY;
+
+    const xpReduced: bigint[] = [];
+
+    v.feePerToken = this._feePerToken(state);
+    for (let i = 0; i < numTokens; i++) {
+      const xpi = xp[i];
+      // if i == tokenIndex, dxExpected = xp[i] * d1 / d0 - newY
+      // else dxExpected = xp[i] - (xp[i] * d1 / d0)
+      // xpReduced[i] -= dxExpected * fee / FEE_DENOMINATOR
+
+      const dxExpected =
+        i === tokenIndex
+          ? (xpi * v.d1) / v.d0 - v.newY
+          : xpi - (xpi * v.d1) / v.d0;
+      xpReduced[i] -= (dxExpected * v.feePerToken) / this.FEE_DENOMINATOR;
```

Comment:
according to the nerve contract this is rather
```suggestion
      const dxExpected =
        i === tokenIndex
          ? (xpi * v.d1) / v.d0 - v.newY
          : xpi - (xpi * v.d1) / v.d0;
      xpReduced[i] = xpi - (dxExpected * v.feePerToken) / this.FEE_DENOMINATOR;
```